pkg/streamingpromql/aggregations/avg.go

```diff
@@ -166,42 +166,46 @@ func (g *AvgAggregationGroup) accumulateHistograms(data types.InstantVectorSerie
 
 	var lastUncopiedHistogram *histogram.FloatHistogram
 
-	for i, p := range data.Histograms {
-		idx := (p.T - start) / interval
-		g.groupSeriesCounts[idx]++
+	for inputIdx, p := range data.Histograms {
+		outputIdx := (p.T - start) / interval
+		g.groupSeriesCounts[outputIdx]++
 
-		if g.histograms[idx] == invalidCombinationOfHistograms {
+		if g.histograms[outputIdx] == invalidCombinationOfHistograms {
 			// We've already seen an invalid combination of histograms at this timestamp. Ignore this point.
 			continue
 		}
 
-		if g.histograms[idx] == nil {
+		if g.histograms[outputIdx] == nil {
+			data.Histograms[inputIdx].H = nil // Ensure the FloatHistogram instance is not reused when the HPoint slice is reused, as we'll retain a reference to the FloatHistogram instance below.
```

Comment:
Note to reviewers: this is the key line in this file, the rest of the diff is just renaming things